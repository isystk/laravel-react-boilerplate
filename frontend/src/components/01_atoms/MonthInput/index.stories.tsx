import { Meta, Story } from '@storybook/react'
import React from 'react'
import MonthInput, { MonthInputProps } from './index'
import { useForm } from 'react-hook-form'

export default {
  title: '01_atoms/MonthInput',
  component: MonthInput,
} as Meta

type FormMonthInputs = {
  yearMonth: string
}

const Template: Story = () => {
  const {
    control,
    register,
    formState: { errors },
  } = useForm<FormMonthInputs>()
  const validate = {
    yearMonth: {
      required: '処理年月を入力してください',
    },
  }
  const props: MonthInputProps = {
    name: 'yearMonth',
    placeholder: '処理年月',
    control,
    register,
    validate,
    errors,
  }
  return (
      <div className="w-56">
        <MonthInput {...props} />
      </div>
  )
}

export const Primary = Template.bind({})
Primary.storyName = 'プライマリ'
Primary.args = {}
