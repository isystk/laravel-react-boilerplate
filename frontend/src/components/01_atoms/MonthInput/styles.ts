import { css, keyframes } from '@emotion/css'

export const monthInput = css``
