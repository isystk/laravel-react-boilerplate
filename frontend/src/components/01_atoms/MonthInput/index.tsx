import React, { FC } from 'react'
import { ContainerProps } from 'types'
import { connect } from '@/components/hoc'
import { Controller } from 'react-hook-form'
import MonthPickerInput from 'react-month-picker-input'
import 'react-month-picker-input/dist/react-month-picker-input.css'
import 'react-datepicker/dist/react-datepicker.css'
import * as styles from './styles'

/** MonthInputProps Props */
export type MonthInputProps = {
  small: boolean
  name
  placeholder: string
  value
  onChange
  control
  register
  validate
  errors
}
/** Presenter Props */
export type PresenterProps = MonthInputProps

const MonthInput: FC<PresenterProps> = ({
  small,
  placeholder,
  value,
  ...props
}) => (
  <MonthPickerInput
    lang="ja"
    // mode="calendarOnly" // 'normal', 'readOnly'
    monthFormat="long"
    i18n={{
      dateFormat: { ja: 'YYYY/MM' },
      monthNames: {
        ja: [
          '1月',
          '2月',
          '3月',
          '4月',
          '5月',
          '6月',
          '7月',
          '8月',
          '9月',
          '10月',
          '11月',
          '12月',
        ],
      },
    }}
    closeOnSelect
    inputProps={{
      placeholder,
      className: `${styles.monthInput} ${
        small ? 'p-2' : 'p-3'
      } bg-gray-200  border border-gray-200 rounded  leading-tight focus:outline-none focus:bg-white focus:border-gray-500`,
    }}
  />
)

/** Presenter Component */
const MonthInputPresenter: FC<PresenterProps> = ({
  name,
  onChange,
  control,
  register,
  validate,
  errors,
  ...props
}) => (
  <>
    {register ? (
      <Controller
        control={control}
        name={name}
        render={({ field: { onChange, value } }) => (
          <>
            <MonthInput
              {...props}
              {...register(name, validate[name])}
              value={value}
              onChange={(maskedValue, selectedYear, selectedMonth) => {
                onChange(maskedValue)
              }}
            />
            {errors[name] && (
              <span className="pt-4 text-red-500">{errors[name].message}</span>
            )}
          </>
        )}
      />
    ) : (
      <MonthInput
        {...props}
        onChange={(maskedValue, selectedYear, selectedMonth) => {
          console.log(maskedValue)
          onChange(maskedValue)
        }}
      />
    )}
  </>
)

/** Container Component */
const MonthInputContainer: React.FC<
  ContainerProps<MonthInputProps, PresenterProps>
> = ({ presenter, ...props }) => {
  return presenter({
    ...props,
  })
}

export default connect<MonthInputProps, PresenterProps>(
  'MonthInput',
  MonthInputPresenter,
  MonthInputContainer
)
