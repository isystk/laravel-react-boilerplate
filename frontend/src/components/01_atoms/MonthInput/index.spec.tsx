import React from 'react'
import renderer from 'react-test-renderer'
import MonthInput, { MonthInputProps } from './index'
import { useForm } from 'react-hook-form'
import { renderHook } from '@testing-library/react-hooks'

type FormMonthInputs = {
  yearMonth: string
}

describe('MonthInput', () => {
  it('Match Snapshot', () => {
    const { result } = renderHook(() => useForm<FormMonthInputs>())

    const {
      control,
      register,
      formState: { errors },
    } = result.current

    const validate = {
      yearMonth: {
        required: '処理年月を入力してください',
      },
    }
    const props: MonthInputProps = {
      name: 'yearMonth',
      placeholder: '処理年月',
      control,
      register,
      validate,
      errors,
    }

    const component = renderer.create(<MonthInput {...props} />)
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
