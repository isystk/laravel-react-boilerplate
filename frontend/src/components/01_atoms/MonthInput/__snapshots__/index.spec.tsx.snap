// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthInput Match Snapshot 1`] = `
<div
  className="react-month-picker"
>
  <input
    className="css-0 p-3 bg-gray-200  border border-gray-200 rounded  leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onPaste={[Function]}
    placeholder="生年月日"
    type="text"
    value=""
  />
</div>
`;
